### DISCUSSION TITLE
Trouble adding random trips to scenario

### DISCUSSION THREAD

**Discussion by pv2-0** [eclipse/mosaic#48] [Q&A] [answered]
Hi,

I am trying to simulate V2X communication in a traffic network and am having difficulty when importing the routes from my rou.xml file to my database. I created my rou.xml file using randomtrips.py and imported them to the scenario according to the instructions [on this page](https://www.eclipse.org/mosaic/docs/building_scenarios/scenario_convert/#importing-routes-to-your-scenario). However, I am getting this message:

> 2:26:26.029 INFO  - initializing database
> 22:26:26.033 INFO  - loading database...
> 22:26:26.536 INFO  - success
> 
> 22:26:26.586 INFO  - no net file for import given, ignoring!
> 22:26:26.590 INFO  - parsing route file
> 22:26:27.356 INFO  - no global route definitions found (probably per vehicle routes?)
> 22:26:27.365 INFO  -
> 
> 22:26:27.368 INFO  - processing vehicle 1/954  
> 22:26:27.386 INFO  - processing vehicle 1/954  edge 1/39
> 22:26:27.389 ERROR |- Sumo edge id '564348420#1' in route '0' not found in database, please check the input file! Skipping!
> 22:26:27.401 WARN  |- vehicle with id 0 has no valid route reference, ignoring
> 22:26:27.401 INFO  - processing vehicle 2/954  
> 22:26:27.402 INFO  - processing vehicle 2/954  edge 1/18
> 22:26:27.402 ERROR |- Sumo edge id '-9266408#5' in route '0' not found in database, please check the input file! Skipping!
> 22:26:27.402 WARN  |- vehicle with id 1 has no valid route reference, ignoring
> 22:26:27.405 INFO  - processing vehicle 3/954  
> 22:26:27.405 INFO  - processing vehicle 3/954  edge 1/14
> 22:26:27.407 ERROR |- Sumo edge id '629278226' in route '0' not found in database, please check the input file! Skipping!
> 22:26:27.408 WARN  |- vehicle with id 2 has no valid route reference, ignoring
> 22:26:27.408 INFO  - processing vehicle 4/954  
> 22:26:27.409 INFO  - processing vehicle 4/954  edge 1/28
> 22:26:27.409 ERROR |- Sumo edge id '755543598#0' in route '0' not found in database, please check the input file! Skipping!
> 22:26:27.417 WARN  |- vehicle with id 3 has no valid route reference, ignoring
> 22:26:27.417 INFO  - processing vehicle 5/954  
> 22:26:27.418 INFO  - processing vehicle 5/954  edge 1/11
> 22:26:27.418 ERROR |- Sumo edge id '234939832#2' in route '0' not found in database, please check the input file! Skipping!
> 22:26:27.418 WARN  |- vehicle with id 4 has no valid route reference, ignoring
> 22:26:27.419 INFO  - processing vehicle 6/954  
> 22:26:27.420 INFO  - processing vehicle 6/954  edge 1/10
> 22:26:27.425 ERROR |- Sumo edge id '-629278230#1' in route '0' not found in database, please check the input file! Skipping!
> 22:26:27.425 WARN  |- vehicle with id 5 has no valid route reference, ignoring
> 22:26:27.428 INFO  - processing vehicle 7/954  
> 22:26:27.429 INFO  - processing vehicle 7/954  edge 1/9
> 22:26:27.429 ERROR |- Sumo edge id '-169926081#11' in route '0' not found in database, please check the input file! Skipping!
> 22:26:27.430 WARN  |- vehicle with id 6 has no valid route reference, ignoring
> 22:26:27.431 INFO  - processing vehicle 8/954  
> 22:26:27.431 INFO  - processing vehicle 8/954  edge 1/23
> 22:26:27.432 ERROR |- Sumo edge id '579586365#2' in route '0' not found in database, please check the input file! Skipping!
> 22:26:27.432 WARN  |- vehicle with id 7 has no valid route reference, ignoring
> 22:26:27.435 INFO  - processing vehicle 9/954  
> 22:26:27.437 INFO  - processing vehicle 9/954  edge 1/25
> 22:26:27.441 ERROR |- Sumo edge id '-629278230#1' in route '0' not found in database, please check the input file! Skipping!
> 22:26:27.442 WARN  |- vehicle with id 8 has no valid route reference, ignoring
> 22:26:27.442 INFO  - processing vehicle 10/954  
> 22:26:27.442 INFO  - processing vehicle 10/954  edge 1/38
> 22:26:27.444 ERROR |- Sumo edge id '630644330#1' in route '0' not found in database, please check the input file! Skipping!
> 22:26:27.444 WARN  |- vehicle with id 9 has no valid route reference, ignoring
> 22:26:27.445 INFO  - processing vehicle 11/954  
> 22:26:27.445 INFO  - processing vehicle 11/954  edge 1/22
> 22:26:27.445 ERROR |- Sumo edge id '630644330#0' in route '0' not found in database, please check the input file! Skipping!
> 22:26:27.446 WARN  |- vehicle with id 10 has no valid route reference, ignoring
> 22:26:27.447 INFO  - processing vehicle 12/954  
> 22:26:27.453 INFO  - processing vehicle 12/954  edge 1/38
> 22:26:27.453 ERROR |- 
> 
> Sumo edge id '169926081#6' in route '0' not found in database, please check the input file! Skipping!

It seems like it is unable to import any routes because when I try to run my scenario, I get the following:

> 2021-02-23 22:45:36,921 INFO  ROOT - Running Eclipse MOSAIC 20.0 on Java JRE v11.0.10 (Ubuntu)
> 2021-02-23 22:45:37,513 INFO  FederationManagement - Start federation with id 'atlanta_dt'
> 2021-02-23 22:45:37,516 INFO  FederationManagement - Add ambassador/federate with id 'application'
> 2021-02-23 22:45:37,518 INFO  FederationManagement - Add ambassador/federate with id 'mapping'
> 2021-02-23 22:45:37,519 INFO  FederationManagement - Add ambassador/federate with id 'sumo'
> 2021-02-23 22:45:37,519 INFO  FederationManagement - Deploying federate 'sumo' locally in ./tmp/sumo
> 2021-02-23 22:45:37,539 INFO  FederationManagement - Starting federate 'sumo' locally in ./tmp/sumo
> 2021-02-23 22:45:37,548 INFO  FederationManagement - Add ambassador/federate with id 'output'
> 22:45:40 - Simulating: 0ns (0.0s / 1000.0s) - 0.0% (RTF:0.00, ETC:unknown)     22:45:42 - Simulating: 0ns (0.0s / 1000.0s) - 0.0% (RTF:0.00, ETC:unknown)                        

>  Stopping simulation due to a critical error:
> - IllegalArgumentException: Unknown route 1 for vehicle with departure time 0
> - Root Cause: IllegalArgumentException: Unknown route 1 for vehicle with departure time 0
>  Please see the log files for details.
> - Log-Directory: logs/log-20210223-224536-atlanta_dt
> 
>  Eclipse MOSAIC will now shutdown.

I would appreciate any advice on how to fix this or if there is a simpler way to generate random trips in MOSAIC altogether. Thank you!

**Comment by kschrab** [comment, marked as answer]
The latest release 20.0 requires to build a SUMO network file from OpenStreetMap using [`scenario-convert`](https://www.dcaiti.tu-berlin.de/research/simulation/download) with the `--osm2db` + `--db2sumo`, or `--osm2sumo`, or `--osm2mosaic` option. From your output it seems that your net-file was not generated using scenario-convert, but directly with netconvert. However, the upcoming release will fix this and will allow you to import any net-file and corresponding route-file you want. You have two options now:
 1. You build your scenario with `scenario-convert` using an OSM file as input. If you use the option `--osm2mosaic -i path/to/map.osm --generate-routes` a fully functioning scenario should be generated with some vehicles driving on some random routes.
 2. You wait until the next release which is scheduled for end of March. With this version you should be able to import any network- and route-file you want.

**Reply by pv2-0** [reply]
Thank you! Once I generate the routes, I still have to spawn vehicles in the mapping_config.json file, right? How do I specify for half of the vehicles to be of one type (ex. not automated) and the other half to be of another type (automated)?  Do I have to spawn vehicles individually and assign each a different route id? It seems like there should be a better way.

**Reply by kschrab** [reply]
Yes, you have to define so called "vehicle spawners" in the mapping, which each generate a constant flow of vehicles. Each flow is assigned to one specific route id, but you can have multiple vehicle types for each spawner, even distribute them unevenly by using different weights. Just have a look into the [Barnim scenario and tutorial](https://www.eclipse.org/mosaic/tutorials/barnim_basic), in which different vehicle types are assigned to one vehicle spawner.
